docs/DeveloperGuide.md

```diff
@@ -2,37 +2,615 @@
+{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the 
+original source as well}
 
-## Design & implementation
+## Design
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+### Architecture
 
+![](.\\uml\\diagrams\\ArchitectureDiagram.png "Architecture Diagram")
+
+The **_Architecture Diagram_** given above explains the high-level design of `CLIAlgo`.
+
+Given below is a quick overview of the main components and how they interact with each other.
+
+#### Main components of the architecture
+
+- `CLIAlgo`: consist of only 1 main method and it is responsible for initializing the componnents
+in the correct sequence and connects them with each other during runtime.
+- `Ui`: Responsible for handling all interactions with the user.
+- `Parser`: Responsible for making sense of all commands entered by the user.
+- `Topic`: A class responsible for managing all files allocated to single topic in CS2040C.
+- `TopicManager`: A class responsible for handling operations involving all `Topic`.
+- `command`: A collection of `Command` objects which handles the different functionalities of `CLIAlgo`.
+- `file`: A collection of files used in CS2040C.
+- `storage`: A collection of classes responsible for reading, storing and writing data to the hard disk.
+
+#### How the architecture components interact with each other
+
+The **_Sequence Diagram_** below shows a high level overview of how the components interact with each other
+
+![](.\\sequence\\diagrams\\Architecture.png "Filter by Topic Sequence Diagram")
+
+### Ui
+**API** : Ui.java
+Here is a class diagram of the `Ui` component which is responsible for handling all interactin with the User.
+
+![](.\\uml\\diagrams\\UiClass.png "Parser Class Diagram")
+
+The `Ui` component:
+- Reads in the full command keyed in by the user for the `Parser`.
+- Prints error messages when the user provides invalid inputs.
+- Prints confirmation message when a command have been executed successfully.
+
+### Parser
+**API** : `Parser.java`
+Here is a class diagram of the `Parser` component which is responsible for processing commands
+and preparing the appropriate `Command` object.
+
+![](.\\uml\\diagrams\\ParserClass.png "Parser Class Diagram")
+
+The `Parser` component:
+- Reads in the full command keyed in by the user through the `Ui` class.
+- Extracts out the keywords from the command such as `command type`, `topic`, `file name` using
+the `StringManipulation` interface.
+- Verify the validity of the user's input command.
+- Handle cases where the user keys in an invalid command.
+- Returns the appropriate `Command` object that will be executed by `CLIAlgo`.
+
+### Storage
+**API** : `FileManager.java`
+
+Here is a class diagram of the `FileManager` which facilitates the storage
+function of the application.
+
+![](.\\uml\\diagrams\\FileManagerClass.png "FileManager Class Diagram")
```

Comment:
I think the diagram's focus would be clearer without the `Scanner` and `FileWriter` classes, as their functions can be described using labels.